README.md
# OSS evolution of classic nx technology - OpenNX CE

All these years, the classic nx was not as dead as it seemed ;)
It is used in production and develops as fast as it can.

I am very grateful to the developers of the [ArcticaProject/nx-libs]
(https://github.com/ArcticaProject/nx-libs) for maintaining backward
compatibility and the opportunity to use their libraries instead
of self-assembly.

Compared to the original opennx new features added by community:

- Building with wxWidgets 3.0.4

- For win32 building with modified !M NX-libs uder cygwin. Xming not used.
Sources of these libs may be found at [old git repo](http://git.etersoft.ru/people/dimbor/packages/?p=NXClient-W32.git)

- CUPS client mode: client's system CUPS used directlty
This need to provide username/pass on each printer (like on win-resources);

- Pulseaudio sound: direct, tunnelled, with or without resampling.
On Windows pulseaudio6 binarys are included;

- Localization of windows sharenames;

- On Windows original fonts from nxclient are included.

- ...

It worked with our own freenx-server
And with classic nxserver 3.x.x compatibility remained.

Debian package home-maded for Devuan ASCII now. There is a suspicion that
under Debian Stretch everything will be fine.

Windows installer can be downloaded at [unixforum.org](https://unixforum.org/up/nxman/OpenNXCE-1.0.0-r15.5.exe)

dimbor. 2019.


ChangeLog
2013-11-25 17:58  Fritz Elfert <felfert@users.sf.net>

	* [r729] osdep.c:
	  Fixed another typo in OSX specific code.

2013-11-25 17:53  Fritz Elfert <felfert@users.sf.net>

	* [r728] packaging/win32/cygnx/getnxsetup:
	  Update URL of old 3.5.0-9 windows setup.

2013-11-25 17:31  Fritz Elfert <felfert@users.sf.net>

	* [r727] osdep.c:
	  Fixed typo in OSX specific code.

2013-11-25 17:22  Fritz Elfert <felfert@users.sf.net>

	* [r726] packaging/win32/pulse/urls.txt:
	  Updated download URL for json-c

2013-11-25 15:59  Fritz Elfert <felfert@users.sf.net>

	* [r725] getnxsrcpkg, opennx.spec.in, opennxApp.cpp, opennxApp.h,
	  osdep.c, packaging/MacOSX/pkgResources/InstallationCheck,
	  po/de.po, po/ja.po, po/opennx.pot, po/ru.po:
	  - Updated fetch URL for nomachine sources.
	  - Support Fedora 19
	  - Support alternate path for quartz-wm on OSX

2012-08-05 17:52  Fritz Elfert <felfert@users.sf.net>

	* [r724] Makefile.am:
	  - Honor RPMBUUILD_OPTS during RPM build

2012-08-01 02:48  Fritz Elfert <felfert@users.sf.net>

	* [r723] packaging/win32/xming/Makefile.am:
	  - Fixed Xming download

2012-08-01 01:35  Fritz Elfert <felfert@users.sf.net>

	* [r722] packaging/win32/Makefile.am:
	  - Quit using isccwrap in mingw build

2012-08-01 01:13  Fritz Elfert <felfert@users.sf.net>

	* [r721] packaging/win32/findrequires.pl.in:
	  - Do case-insensitive compare while searching for dependencies in
	  mingw build

2012-07-31 23:52  Fritz Elfert <felfert@users.sf.net>

	* [r720] packaging/win32/cygnx/Makefile.am:
	  - Fixed fetch of unnounp in nxwin build variant

2012-07-31 13:51  Fritz Elfert <felfert@users.sf.net>

	* [r719] packaging/win32/pulse/patches/libsndfile-win32.patch:
	  - Added patch for building libsndfile on newer mingw

2012-07-31 04:57  Fritz Elfert <felfert@users.sf.net>

	* [r718] packaging/win32/pulse/Makefile.am:
	  - Fix build of json-c using latest F17 mingw32-configure

2012-07-31 02:57  Fritz Elfert <felfert@users.sf.net>

	* [r717] packaging/win32/pulse/Makefile.am:
	  - Fixed minor glitch in makefile

2012-07-31 02:26  Fritz Elfert <felfert@users.sf.net>

	* [r716] Makefile.am:
	  - Build bin2hdr helper always static

2012-07-31 01:22  Fritz Elfert <felfert@users.sf.net>

	* [r715] configure.ac:
	  - Use iscc package from OBS as default for mingw cross builds.

2012-07-30 18:53  Fritz Elfert <felfert@users.sf.net>

	* [r714] packaging/MacOSX/Makefile.am:
	  - OSX-build: Invoke sudo during clean-local only when necessary

2012-07-02 15:56  Fritz Elfert <felfert@users.sf.net>

	* [r713] opennx.spec.in, opennxApp.cpp, opennxApp.h:
	  - Fix RPM postinstall: create symlink for libpulse
	  - Disable any GUI dialogs while performing --exportres

2012-06-19 10:45  Fritz Elfert <felfert@users.sf.net>

	* [r712] po/LINGUAS, po/ja.po:
	  - Added japanese translation stubs.

2012-06-02 21:35  Fritz Elfert <felfert@users.sf.net>

	* [r711] m4/compat.m4:
	  - Added missing m4 file

2012-06-02 19:18  Fritz Elfert <felfert@users.sf.net>

	* [r710] configure.ac, m4/flibcurl.m4,
	  patches/nxcomp-zz_png15.patch, po/de.po:
	  - Added patch for building against latest libpng on Fedora 17

2012-05-12 22:06  Fritz Elfert <felfert@users.sf.net>

	* [r709] UsbIp.h, mylog.h:
	  - Fixed compile error on Ubuntu 12.04 due to patched wxWidgets

2012-03-23 18:18  Fritz Elfert <felfert@users.sf.net>

	* [r708] UnixImageSettingsDialog.cpp, UnixImageSettingsDialog.h,
	  VncImageSettingsDialog.cpp, opennx.pjd, res/opennx.xrc:
	  - Prevent initial focus from selecting first radiobutton in
	  custom image settings.

2012-03-22 13:53  Fritz Elfert <felfert@users.sf.net>

	* [r707] ShareProperties.cpp, ShareProperties.h,
	  ShareProperties_symbols.h, opennx.pjd,
	  packaging/win32/xming/patches/xming-kbhook.patch, res/opennx.xrc:
	  - Fixed bug #3509691
	  - Added missing kbhook in Xming

2012-03-19 10:15  Fritz Elfert <felfert@users.sf.net>

	* [r706] packaging/win32/xming/Makefile.am,
	  packaging/win32/xming/patches/xming-kioskhack.patch:
	  - Added abiltiy to disable the NX magic pixel in fullscreen mode
	  (win32)

2012-03-16 13:01  Fritz Elfert <felfert@users.sf.net>

	* [r705] MySession.cpp, SessionProperties.cpp, SessionProperties.h,
	  SessionProperties_symbols.h, opennx.pjd,
	  patches/nxcomp-zz_debugfix.patch,
	  patches/nxcomp-zz_kiosk-hack.patch, po/de.po, po/opennx.pot,
	  po/ru.po, res/opennx.xrc:
	  - Fixed german translation (Bugs 3496973 and 3497020).
	  - Fixed build of nxcomp with INFO logging enabled.
	  - Added abiltiy to disable the NX magic pixel in fullscreen mode
	  (not yet working on Windows).

2012-03-14 14:55  Fritz Elfert <felfert@users.sf.net>

	* [r704] MySession.cpp, MySession.h, MyXmlConfig.cpp:
	  - Added workaround for server bug: attach to shadow session
	  fails, if server's DisplayBase is >= 10000
	  - Simplified ssh logging

2012-03-12 16:45  Fritz Elfert <felfert@users.sf.net>

	* [r703] debian/postinst:
	  - Added missing symlink for libpulse in debian postinst script

2012-02-24 11:41  Fritz Elfert <felfert@users.sf.net>

	* [r702] debian/rules, tracelogApp.cpp:
	  - Fixed remote ip regex in tracelog
	  - Fixed debian build

2012-02-22 14:22  Fritz Elfert <felfert@users.sf.net>

	* [r701] debian/rules:
	  - Fixed build rule for debian

2012-02-22 11:56  Fritz Elfert <felfert@users.sf.net>

	* [r700] MyIPC.cpp, MyIPC.h:
	  - Added missing error handling for expired evaluation license on
	  commercial NX server.

2012-02-22 10:14  Fritz Elfert <felfert@users.sf.net>

	* [r699] debian/rules:
	  - Fix debian build

2012-02-22 04:21  Fritz Elfert <felfert@users.sf.net>

	* [r698] SessionProperties.cpp, debian/rules,
	  packaging/win32/xming/Makefile.am,
	  packaging/win32/xming/patches/xming-kbhook.patch,
	  patches/nxcomp-irlimit.patch, po/de.po, po/opennx.pot, po/ru.po:
	  - Fixed keyboard grabbing on Windows
	  - Added missing irlimit patch
	  - Fixed debian rules

2012-02-21 09:50  Fritz Elfert <felfert@users.sf.net>

	* [r697] debian/rules, opennx.spec.in:

md5.c
/*
  Copyright (C) 1999 Aladdin Enterprises.  All rights reserved.

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  L. Peter Deutsch
  ghost@aladdin.com

 */
/*$Id: md5.c 24 2006-12-06 03:21:20Z felfert $ */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321.
  It is derived directly from the text of the RFC and not from the
  reference implementation.

  The original and principal author of md5.c is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5).
  1999-05-03 lpd Original version.
 */


#include "md5.h"
#include <string.h>

#ifdef TEST
/*
 * Compile with -DTEST to create a self-contained executable test program.
 * The test program should print out the same values as given in section
 * A.5 of RFC 1321, reproduced below.
 */

main()
{
    static const char *const test[7] = {
	"", /*d41d8cd98f00b204e9800998ecf8427e*/
	"a", /*0cc175b9c0f1b6a831c399e269772661*/
	"abc", /*900150983cd24fb0d6963f7d28e17f72*/
	"message digest", /*f96b697d7cb7938d525a2f31aaf161d0*/
	"abcdefghijklmnopqrstuvwxyz", /*c3fcd3d76192e4007dfb496cca67e13b*/
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789",
				/*d174ab98d277d9f5a5611c2c9f419d9f*/
	"12345678901234567890123456789012345678901234567890123456789012345678901234567890" /*57edf4a22be3c955ac49da2e2107b67a*/
    };
    int i;

    for (i = 0; i < 7; ++i) {
	md5_state_t state;
	md5_byte_t digest[16];
	int di;

	md5_init(&state);
	md5_append(&state, (const md5_byte_t *)test[i], strlen(test[i]));
	md5_finish(&state, digest);
	printf("MD5 (\"%s\") = ", test[i]);
	for (di = 0; di < 16; ++di)
	    printf("%02x", digest[di]);
	printf("\n");
    }
    return 0;
}
#endif /* TEST */


/*
 * For reference, here is the program that computed the T values.
 */
#if 0
#include <math.h>
main()
{
    int i;
    for (i = 1; i <= 64; ++i) {
	unsigned long v = (unsigned long)(4294967296.0 * fabs(sin((double)i)));
	printf("#define T%d 0x%08lx\n", i, v);
    }
    return 0;
}
#endif
/*
 * End of T computation program.
 */
#define T1 0xd76aa478
#define T2 0xe8c7b756
#define T3 0x242070db
#define T4 0xc1bdceee
#define T5 0xf57c0faf
#define T6 0x4787c62a
#define T7 0xa8304613
#define T8 0xfd469501
#define T9 0x698098d8
#define T10 0x8b44f7af
#define T11 0xffff5bb1
#define T12 0x895cd7be
#define T13 0x6b901122
#define T14 0xfd987193
#define T15 0xa679438e
#define T16 0x49b40821
#define T17 0xf61e2562
#define T18 0xc040b340
#define T19 0x265e5a51
#define T20 0xe9b6c7aa
#define T21 0xd62f105d
#define T22 0x02441453
#define T23 0xd8a1e681
#define T24 0xe7d3fbc8
#define T25 0x21e1cde6
#define T26 0xc33707d6
#define T27 0xf4d50d87
#define T28 0x455a14ed
#define T29 0xa9e3e905
#define T30 0xfcefa3f8
#define T31 0x676f02d9
#define T32 0x8d2a4c8a
#define T33 0xfffa3942
#define T34 0x8771f681
#define T35 0x6d9d6122
#define T36 0xfde5380c
#define T37 0xa4beea44
#define T38 0x4bdecfa9
#define T39 0xf6bb4b60
#define T40 0xbebfbc70
#define T41 0x289b7ec6
#define T42 0xeaa127fa
#define T43 0xd4ef3085
#define T44 0x04881d05
#define T45 0xd9d4d039
#define T46 0xe6db99e5
#define T47 0x1fa27cf8
#define T48 0xc4ac5665
#define T49 0xf4292244
#define T50 0x432aff97
#define T51 0xab9423a7
#define T52 0xfc93a039
#define T53 0x655b59c3
#define T54 0x8f0ccc92
#define T55 0xffeff47d
#define T56 0x85845dd1
#define T57 0x6fa87e4f
#define T58 0xfe2ce6e0
#define T59 0xa3014314
#define T60 0x4e0811a1
#define T61 0xf7537e82
#define T62 0xbd3af235
#define T63 0x2ad7d2bb
#define T64 0xeb86d391

static void
md5_process(md5_state_t *pms, const md5_byte_t *data /*[64]*/)
{
    md5_word_t
	a = pms->abcd[0], b = pms->abcd[1],
	c = pms->abcd[2], d = pms->abcd[3];
    md5_word_t t;

#ifndef ARCH_IS_BIG_ENDIAN
# define ARCH_IS_BIG_ENDIAN 1	/* slower, default implementation */
#endif
#if ARCH_IS_BIG_ENDIAN

    /*
     * On big-endian machines, we must arrange the bytes in the right
     * order.  (This also works on machines of unknown byte order.)
     */
    md5_word_t X[16];
    const md5_byte_t *xp = data;
    int i;

    for (i = 0; i < 16; ++i, xp += 4)
	X[i] = xp[0] + (xp[1] << 8) + (xp[2] << 16) + (xp[3] << 24);

#else  /* !ARCH_IS_BIG_ENDIAN */

    /*
     * On little-endian machines, we can process properly aligned data
     * without copying it.
     */
    md5_word_t xbuf[16];
    const md5_word_t *X;

    if (!((data - (const md5_byte_t *)0) & 3)) {
	/* data are properly aligned */
	X = (const md5_word_t *)data;
    } else {
	/* not aligned */
	memcpy(xbuf, data, 64);
	X = xbuf;
    }
#endif

#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

    /* Round 1. */
    /* Let [abcd k s i] denote the operation
       a = b + ((a + F(b,c,d) + X[k] + T[i]) <<< s). */
#define F(x, y, z) (((x) & (y)) | (~(x) & (z)))
#define SET(a, b, c, d, k, s, Ti)  t = a + F(b,c,d) + X[k] + Ti;  a = ROTATE_LEFT(t, s) + b
    /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  7,  T1);
    SET(d, a, b, c,  1, 12,  T2);
    SET(c, d, a, b,  2, 17,  T3);
    SET(b, c, d, a,  3, 22,  T4);
    SET(a, b, c, d,  4,  7,  T5);
    SET(d, a, b, c,  5, 12,  T6);
    SET(c, d, a, b,  6, 17,  T7);
    SET(b, c, d, a,  7, 22,  T8);
    SET(a, b, c, d,  8,  7,  T9);
    SET(d, a, b, c,  9, 12, T10);
    SET(c, d, a, b, 10, 17, T11);
    SET(b, c, d, a, 11, 22, T12);
    SET(a, b, c, d, 12,  7, T13);
    SET(d, a, b, c, 13, 12, T14);
    SET(c, d, a, b, 14, 17, T15);
    SET(b, c, d, a, 15, 22, T16);
#undef SET

     /* Round 2. */
     /* Let [abcd k s i] denote the operation
          a = b + ((a + G(b,c,d) + X[k] + T[i]) <<< s). */
#define G(x, y, z) (((x) & (z)) | ((y) & ~(z)))
#define SET(a, b, c, d, k, s, Ti)   t = a + G(b,c,d) + X[k] + Ti;   a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  1,  5, T17);
    SET(d, a, b, c,  6,  9, T18);
    SET(c, d, a, b, 11, 14, T19);
    SET(b, c, d, a,  0, 20, T20);
    SET(a, b, c, d,  5,  5, T21);
    SET(d, a, b, c, 10,  9, T22);
    SET(c, d, a, b, 15, 14, T23);
    SET(b, c, d, a,  4, 20, T24);
    SET(a, b, c, d,  9,  5, T25);
    SET(d, a, b, c, 14,  9, T26);
    SET(c, d, a, b,  3, 14, T27);
    SET(b, c, d, a,  8, 20, T28);
    SET(a, b, c, d, 13,  5, T29);
    SET(d, a, b, c,  2,  9, T30);
    SET(c, d, a, b,  7, 14, T31);
    SET(b, c, d, a, 12, 20, T32);
#undef SET

     /* Round 3. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + H(b,c,d) + X[k] + T[i]) <<< s). */
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define SET(a, b, c, d, k, s, Ti)   t = a + H(b,c,d) + X[k] + Ti;   a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  5,  4, T33);
    SET(d, a, b, c,  8, 11, T34);
    SET(c, d, a, b, 11, 16, T35);
    SET(b, c, d, a, 14, 23, T36);
    SET(a, b, c, d,  1,  4, T37);
    SET(d, a, b, c,  4, 11, T38);
    SET(c, d, a, b,  7, 16, T39);
    SET(b, c, d, a, 10, 23, T40);
    SET(a, b, c, d, 13,  4, T41);
    SET(d, a, b, c,  0, 11, T42);
    SET(c, d, a, b,  3, 16, T43);
    SET(b, c, d, a,  6, 23, T44);
    SET(a, b, c, d,  9,  4, T45);
    SET(d, a, b, c, 12, 11, T46);
    SET(c, d, a, b, 15, 16, T47);
    SET(b, c, d, a,  2, 23, T48);
#undef SET

     /* Round 4. */
     /* Let [abcd k s t] denote the operation
          a = b + ((a + I(b,c,d) + X[k] + T[i]) <<< s). */
#define I(x, y, z) ((y) ^ ((x) | ~(z)))
#define SET(a, b, c, d, k, s, Ti) t = a + I(b,c,d) + X[k] + Ti; a = ROTATE_LEFT(t, s) + b
     /* Do the following 16 operations. */
    SET(a, b, c, d,  0,  6, T49);
    SET(d, a, b, c,  7, 10, T50);
    SET(c, d, a, b, 14, 15, T51);
    SET(b, c, d, a,  5, 21, T52);
    SET(a, b, c, d, 12,  6, T53);
    SET(d, a, b, c,  3, 10, T54);
    SET(c, d, a, b, 10, 15, T55);
    SET(b, c, d, a,  1, 21, T56);
    SET(a, b, c, d,  8,  6, T57);
    SET(d, a, b, c, 15, 10, T58);
    SET(c, d, a, b,  6, 15, T59);
    SET(b, c, d, a, 13, 21, T60);
    SET(a, b, c, d,  4,  6, T61);
    SET(d, a, b, c, 11, 10, T62);
    SET(c, d, a, b,  2, 15, T63);
    SET(b, c, d, a,  9, 21, T64);
#undef SET

     /* Then perform the following additions. (That is increment each
        of the four registers by the value it had before this block
        was started.) */
    pms->abcd[0] += a;
    pms->abcd[1] += b;
    pms->abcd[2] += c;
    pms->abcd[3] += d;
}

void
md5_init(md5_state_t *pms)
{
    pms->count[0] = pms->count[1] = 0;
    pms->abcd[0] = 0x67452301;
    pms->abcd[1] = 0xefcdab89;
    pms->abcd[2] = 0x98badcfe;
    pms->abcd[3] = 0x10325476;
}

void
md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes)
{
    const md5_byte_t *p = data;
    int left = nbytes;
    int offset = (pms->count[0] >> 3) & 63;
    md5_word_t nbits = (md5_word_t)(nbytes << 3);

    if (nbytes <= 0)
	return;

    /* Update the message length. */
    pms->count[1] += nbytes >> 29;
    pms->count[0] += nbits;
    if (pms->count[0] < nbits)
	pms->count[1]++;

    /* Process an initial partial block. */
    if (offset) {
	int copy = (offset + nbytes > 64 ? 64 - offset : nbytes);

	memcpy(pms->buf + offset, p, copy);
	if (offset + copy < 64)
	    return;
	p += copy;
	left -= copy;
	md5_process(pms, pms->buf);
    }

    /* Process full blocks. */
    for (; left >= 64; p += 64, left -= 64)
	md5_process(pms, p);

    /* Process a final partial block. */
    if (left)
	memcpy(pms->buf, p, left);
}

void
md5_finish(md5_state_t *pms, md5_byte_t digest[16])
{
    static const md5_byte_t pad[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };
    md5_byte_t data[8];
    int i;

    /* Save the length before padding. */
    for (i = 0; i < 8; ++i)
	data[i] = (md5_byte_t)(pms->count[i >> 2] >> ((i & 3) << 3));
    /* Pad to 56 bytes mod 64. */
    md5_append(pms, pad, ((55 - (pms->count[0] >> 3)) & 63) + 1);
    /* Append the length. */
    md5_append(pms, data, 8);
    for (i = 0; i < 16; ++i)
	digest[i] = (md5_byte_t)(pms->abcd[i >> 2] >> ((i & 3) << 3));
}
